gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/sql/execute/SnappyActivation.java

```diff
@@ -311,13 +311,17 @@ public void setResultDescription(GenericResultDescription resultDescription) {
 
   public static void executeOnLeadNode(SnappySelectResultSet rs, GfxdResultCollector<Object> rc,
       String sql, boolean enableStreaming, long connId, String schema, ParameterValueSet pvs,
-      boolean isPreparedStatement, boolean isUpdateOrDelete)
+      boolean isPreparedStatement, boolean isUpdateOrDelete, LanguageConnectionContext lcc)
       throws StandardException {
     // TODO: KN probably username, statement id and connId to be sent in
     // execution and of course tx id when transaction will be supported.
     LeadNodeExecutionContext ctx = new LeadNodeExecutionContext(connId);
     LeadNodeExecutorMsg msg = new LeadNodeExecutorMsg(sql, schema, ctx, rc, pvs,
         isPreparedStatement, false, isUpdateOrDelete);
+    // release all locks before sending the message else it can lead to deadlocks
+    if (lcc != null) {
+      lcc.getTransactionExecute().releaseAllLocks(true, true);
```

Comment:
@vivekwiz DataDictionary and table read locks that lead to deadlocks on routing of DDLs.

> does some DMLs and selects, that are mixed and both routing and non-routing, will that have no effect with released locks?

Not unless the two operations are part of same transaction. Even when we do have driver initiated transactions, the lead will take care of locks for the routed statements while only non-routed ones will take locks here.

Going forward the plan is to remove these DataDictionary/table locks completely from queries/DMLs which have caused much grief (ResultSet not closed/consumed causing hangs etc). See https://jira.snappydata.io/browse/SNAP-1856